src/PregWrapper.php

```diff
@@ -0,0 +1,175 @@
+        if (false !== $result) {
+            return $result;
+        }
+
+        $result = preg_match_all(self::removeUtf8Modifier($pattern), $subject, $matches, $flags, $offset);
+        if (false !== $result) {
+            return $result;
+        }
+
+        throw new \RuntimeException('Error occurred when calling preg_match_all.', preg_last_error());
+    }
+
+    /**
+     * @param string|string[] $pattern
+     * @param string|string[] $replacement
+     * @param string|string[] $subject
+     * @param int             $limit
+     * @param int             $count
+     *
+     * @throws \RuntimeException
+     *
+     * @return string|string[]
+     */
+    public static function replace($pattern, $replacement, $subject, $limit = -1, &$count = null)
+    {
+        $result = @preg_replace(self::addUtf8Modifier($pattern), $replacement, $subject, $limit, $count);
+        if (null !== $result) {
+            return $result;
+        }
+
+        $result = preg_replace(self::removeUtf8Modifier($pattern), $replacement, $subject, $limit, $count);
+        if (null !== $result) {
+            return $result;
+        }
+
+        throw new \RuntimeException('Error occurred when calling preg_replace.', preg_last_error());
+    }
+
+    /**
+     * @param string|string[] $pattern
+     * @param callback        $callback
+     * @param string|string[] $subject
+     * @param int             $limit
+     * @param int             $count
+     *
+     * @throws \RuntimeException
+     *
+     * @return string|string[]
+     */
+    public static function replaceCallback($pattern, $callback, $subject, $limit = -1, &$count = null)
+    {
+        $result = @preg_replace_callback(self::addUtf8Modifier($pattern), $callback, $subject, $limit, $count);
+        if (null !== $result) {
+            return $result;
+        }
+
+        $result = preg_replace_callback(self::removeUtf8Modifier($pattern), $callback, $subject, $limit, $count);
+        if (null !== $result) {
+            return $result;
+        }
+
+        throw new \RuntimeException('Error occurred when calling preg_replace_callback.', preg_last_error());
+    }
+
+    /**
+     * @param string $pattern
+     * @param string $subject
+     * @param int    $limit
+     * @param int    $flags
+     *
+     * @return string[]
+     */
+    public static function split($pattern, $subject, $limit = -1, $flags = 0)
+    {
+        $result = @preg_split(self::addUtf8Modifier($pattern), $subject, $limit, $flags);
+        if (false !== $result) {
+            return $result;
+        }
+
+        return preg_split(self::removeUtf8Modifier($pattern), $subject, $limit, $flags);
+    }
+
+    /**
+     * @param string|string[] $pattern
+     *
+     * @return string|string[]
+     */
+    private static function addUtf8Modifier($pattern)
+    {
+        if (is_array($pattern)) {
+            return array_map('self::'.__FUNCTION__, $pattern);
```

Comment:
Could be `__METHOD__` instead of `'self::'.__FUNCTION__` I think. Same in `removeUtf8Modifier()`.